hubblestack/extmods/audit/bexpr.py

```diff
@@ -0,0 +1,197 @@
+    """
+    log.debug('Executing bexpr module for check-id: %s' % (check_id))
+    keyword_list = ['AND', 'OR', 'NOT', '(', ')']
+    expression = audit_check['expr'].strip()
+    #Seperating keywords on the basis of space
+    expression = expression.replace('(', ' ( ')
+    expression = expression.replace(')', ' ) ')
+    #Parse the expression and evaluate the result
+    #Splitting the expression on the basis of spaces
+    expr_list = expression.split(" ")
+    expr_list = list(filter(None, expr_list)) #Filtering out empty spaces
+    referred_checks_list = []
+    referred_checks_result = {}
+    for expr in expr_list:
+        if expr.upper() not in keyword_list:
+            referred_checks_list.append(expr)
+
+    #Fetch the result of check from result list and store the result of referred checks
+    error = {}
+    for check in referred_checks_list:
+        check_found = False
+        for result in result_list:
+            if result.get('check_id', '') == check:
+                check_found = True
+                check_result =  result.get('check_result', '')
+                if check_result == "Success":
+                    referred_checks_result[check] = "True"
+                elif check_result == "Failure":
+                    referred_checks_result[check] = "False"
+                else:
+                    error[check] = "Unable to fetch result of referred check due to errors. Check: %s" % (check)
+                break
+        if not check_found:
+            error[check] = "Referred check: %s is not available. Please verify correct check is referred." % (check)
+    if error:
+        raise AuditCheckValidationError(error)
+
+    #Convert the expression now in the format to be parsed by pyparsing module
+    parsed_list = []
+    for expr in expr_list:
+        if expr.upper() not in keyword_list:
+            #Check reference is passed. Pass the fetched value instead of original check id
+            parsed_list.append(referred_checks_result.get(expr))
+        else:
+            parsed_list.append(expr.upper())
+
+    parsed_expr = " ".join(parsed_list)
+
+    #Logic to use boolean expression parser using pyparsing library
+    #We are passing the boolean expression in the following form:
+    # check1 and not (check2 or (check3 and not check4) )
+    # check1 and not ( check2 or ( check3 and not check4 )  )
+    # True and not ( False or ( True and not False ) )
+    ParserElement.enablePackrat()
+
+    TRUE = Keyword("True")
+    FALSE = Keyword("False")
+    boolOperand = TRUE | FALSE | Word(alphas, max=1)
+    boolOperand.setParseAction(BoolOperand)
+
+    boolExpr = infixNotation(
+        boolOperand,
+        [
+            ("NOT", 1, opAssoc.RIGHT, BoolNot),
+            ("AND", 2, opAssoc.LEFT, BoolAnd),
+            ("OR", 2, opAssoc.LEFT, BoolOr),
+        ],
+    )
+    result = boolExpr.parseString(parsed_expr)[0]
+    if not bool(result):
+        return {"result": False,
```

Comment:
We need to put in the logs, which check is pass/fail.

Reply:
Added more detailed failure reason for boolean expressions

Reply:
Added more detailed failure reason for boolean expression